### PR TITLE
[KEP-3203] Fetch and Render CVE JSON Feed

### PR DESCRIPTION
**Pull request by PushkarJ**
## Description

- Pull JSON blob from queried issues
- Use layout output formats + templates to generate HTML table and JSON blob
- Add localized strings and caption for CVE feed
- Add a new page to describe details about CVE feed and how to use it
- Update existing pages and link the official CVE feed from it

### Why are we regenerating JSON blob from the JSON blob in the bucket ?
- Not all fields stored in the bucket are needed 
- JSON download directly to a file with Hugo data templates is not possible
- Gives us better control and ability to add extra fields as needed
- Pulling data through hugo data templates allow us to reuse built-in caching mechanisms

### Why not version control the JSON blob?

To avoid multiple sources of truth. The content in the bucket will remain the only source of truth as the JSON blob is constructed from GitHub issues with label `official-cve-feed` (See this for more info: https://github.com/kubernetes/test-infra/pull/23428). However, as a break glass, there is an option to temporarily push the JSON blob by overriding the layouts json file with raw JSON in cases where push to bucket fails or if bucket is inaccessible. 

### Are the locations for the layouts files correct?

Current locations are mostly my guess, so happy to refactor them as needed

TODO:

- [x] Update JSON blob location to k8s GCS bucket
- [x] Add more description in the PR
- [x] Address PR feedback from superseded PR

Preview: https://deploy-preview-35228--kubernetes-io-main-staging.netlify.app/docs/reference/issues-security/official-cve-feed/
KEP tracker: https://github.com/kubernetes/enhancements/issues/3203
xref: https://github.com/kubernetes/sig-security/issues/1
supersedes https://github.com/kubernetes/website/pull/31051 and https://github.com/kubernetes/website/pull/34765
/hold
/sig security docs

### CODE REVIEW

File: layouts/shortcodes/issues-security.html
Comment (sftim) [inline layouts/shortcodes/issues-security.html:2]: This should be a localized string, really.

**Review by sftim — COMMENTED** [review]
- Paths within `static` that are actually generated should include `generated` in their path. A `find` and `grep` should show some existing paths that follow that convention.
- I'd name `gen-cve-feed.sh` as `fetch-cve-feed.sh`

:question: How do we make sure that people don't accidentally commit that fetched file?

:question: For a container build, shouldn't the fetch happen _inside_ the container?

File: Makefile
Comment (sftim) [inline Makefile:24]: Hugo can fetch this JSON file itself - see https://gohugo.io/templates/data-templates/

I think I'd prefer that, so we're not modifying the website source during a build.

File: layouts/shortcodes/issues-security.html
Comment (sftim) [inline layouts/shortcodes/issues-security.html:11]: This could be a `getJSON` doing a HTTP (well, HTTPS) fetch. Hugo will cache the result.

File: layouts/shortcodes/issues-security.html
Comment (sftim) [reply layouts/shortcodes/issues-security.html:2]: Oh, also try to avoid putting CSS here. Instead, set a style rule that uses a `table.security-cves caption` selector.

File: layouts/shortcodes/issues-security.html
Comment (sftim) [inline layouts/shortcodes/issues-security.html:14]: (nit) Mark this text as English - it's not localized.
```suggestion
    <td lang="en">{{ .summary | htmlEscape }}</td>
```

File: layouts/shortcodes/issues-security.html
Comment (PushkarJ) [reply layouts/shortcodes/issues-security.html]: Updated

File: Makefile
Comment (PushkarJ) [reply Makefile]: Yes good call. It took some work, but got this working without build changes

**Comment by PushkarJ** [comment]
@sftim Thank you for the excellent feedback! I think the [example](https://www.raymondcamden.com/2017/05/18/creating-a-json-feed-for-hugo) you shared on slack, really helped me understand what and how to do this. 

Since with these changes there are no build changes needed. I squashed and rebased the commits so please review this as _almost_ a fresh commit / PR

**Review by nehaLohia27 — APPROVED** [review]

File: content/en/docs/reference/issues-security/official-cve-feed.md
Comment (sftim) [inline content/en/docs/reference/issues-security/official-cve-feed.md:21]: After alpha, it'd be nice to sort the table that this renders.

File: layouts/_default/official-cve-feed.json
Comment (sftim) [inline layouts/_default/official-cve-feed.json:9]: This doesn't look right for the final piece. Should we mark this PR as a work-in-progress?

File: layouts/shortcodes/official-cve-feed.html
Comment (sftim) [inline layouts/shortcodes/official-cve-feed.html:11]: Ideally, we'd make this a site parameter (see the top-level `config.toml`) so that we can change / configure it in one place.

File: content/en/docs/reference/issues-security/official-cve-feed.md
Comment (sftim) [inline content/en/docs/reference/issues-security/official-cve-feed.md:7]: I recommend _not_ using the same name for the shortcode and the layout. For example, name the layout `cve-feed` instead.

File: content/en/docs/reference/issues-security/official-cve-feed.md
Comment (sftim) [inline content/en/docs/reference/issues-security/official-cve-feed.md:13]:
```suggestion
the Kubernetes Security Response Committee. See
[Kubernetes Security and Disclosure Information](/docs/reference/issues-security/security/)
for more details.
```

File: content/en/docs/reference/issues-security/official-cve-feed.md
Comment (sftim) [inline content/en/docs/reference/issues-security/official-cve-feed.md:15]:
```suggestion
The Kubernetes project publishes a programmatically accessible
[JSON Feed](/docs/reference/issues-security/official-cve-feed/index.json) of published security issues.
```

File: content/en/docs/reference/issues-security/official-cve-feed.md
Comment (sftim) [inline content/en/docs/reference/issues-security/official-cve-feed.md:18]: I'd write
```shell
curl -v https://k8s.io/docs/reference/issues-security/official-cve-feed/index.json
```
Also, don't indent these lines.

File: content/en/docs/reference/issues-security/official-cve-feed.md
Comment (sftim) [inline content/en/docs/reference/issues-security/official-cve-feed.md:31]:
```suggestion
The source of truth of this feed is a set of GitHub Issues, filtered by a controlled and
```

**Review by sftim — COMMENTED** [review]
This is great!

Some suggestions. But still, this is great!

**Comment by sftim** [comment]
/retitle [WIP] [KEP-3203] Fetch and Render CVE JSON feed

File: content/en/docs/reference/issues-security/issues.md
Comment (sftim) [inline content/en/docs/reference/issues-security/issues.md:11]:
```suggestion
* Official [list of known CVEs](/docs/reference/issues-security/official-cve-feed/)
  (security vulnerabilities) that have been announced by the
  [Security Response Committee](https://github.com/kubernetes/committee-security-response)
```
?

The key change here is to use a site-relative path rather than `./../official-cve-feed`

File: layouts/_default/official-cve-feed.json
Comment (sftim) [reply layouts/_default/official-cve-feed.json:9]: Would `getJSON "URL_GOES_HERE" | jsonify` also work?

File: layouts/_default/official-cve-feed.json
Comment (sftim) [inline layouts/_default/official-cve-feed.json:12]: Another option:
```suggestion
        {{ T "cve_id" | jsonify }}: {{ .cve_id | jsonify }},
```
If that works how I think it would, repeat the same approach for the other lines.

File: content/en/docs/reference/issues-security/issues.md
Comment (PushkarJ) [reply content/en/docs/reference/issues-security/issues.md:11]: I always miss this :( Thanks for catching it

File: content/en/docs/reference/issues-security/official-cve-feed.md
Comment (PushkarJ) [reply content/en/docs/reference/issues-security/official-cve-feed.md:7]: Done

File: content/en/docs/reference/issues-security/official-cve-feed.md
Comment (PushkarJ) [reply content/en/docs/reference/issues-security/official-cve-feed.md]: Done

File: layouts/_default/official-cve-feed.json
Comment (PushkarJ) [reply layouts/_default/official-cve-feed.json]: Yes this works. Updated

File: layouts/_default/official-cve-feed.json
Comment (PushkarJ) [reply layouts/_default/official-cve-feed.json]: Yes this bucket creation is pending: https://github.com/kubernetes/k8s.io/pull/4009

On the other comment, I am curious what would be the benefit of adding `| jsonify` to the line? 

If I add it it is failing with this error:
`Rebuild failed: failed to render pages: render of "page" failed: "/src/layouts/_default/cve-feed.json:9:115": execute of template failed at <jsonify>: range can’t iterate over `

File: content/en/docs/reference/issues-security/official-cve-feed.md
Comment (PushkarJ) [reply content/en/docs/reference/issues-security/official-cve-feed.md]: Yes agreed. I have added it in future scope here https://github.com/kubernetes/sig-security/issues/1

File: layouts/shortcodes/official-cve-feed.html
Comment (PushkarJ) [reply layouts/shortcodes/official-cve-feed.html]: Added. It does look cleaner

**Comment by PushkarJ** [comment]
/label tide/merge-method-squash

(@sftim All the working suggestions are incorporated in this commit: https://github.com/kubernetes/website/pull/35228/commits/a3ea562a87bebf1adea2835e250e8c8556365b06)

File: config.toml
Comment (PushkarJ) [inline config.toml:173]: TODO: This needs an update before merging

**Comment by PushkarJ** [comment]
/retitle [KEP-3203] Fetch and Render CVE JSON Feed

(The bucket used is official one and has the recent data)

**Comment by PushkarJ** [comment]
/hold cancel

(Since bucket has data already)

File: layouts/shortcodes/cve-feed.html
Comment (sftim) [inline layouts/shortcodes/cve-feed.html:14]: These might need to be:
```suggestion
   <td><a href="{{ .cve_url }}">{{ .cve_id | htmlEscape | safeHTML }}</a></td>
   <td>{{ .summary | htmlEscape | safeHTML }}</td>
```

Look for `&gt;` and `&lt;` in the preview. Marking these escaped strings as safe should fix that (untested).

File: layouts/shortcodes/cve-feed.html
Comment (PushkarJ) [inline layouts/shortcodes/cve-feed.html:14]: @sftim Adding `safeHTML` did the trick here

File: layouts/_default/cve-feed.json
Comment (PushkarJ) [inline layouts/_default/cve-feed.json:20]: @sftim but adding it here did not fix things. Turns out it is a known issue: https://github.com/gohugoio/hugo/issues/7229 so I used the solution described in the issue

Comment (sftim) [reply layouts/_default/cve-feed.json:20]: I recommend adding a (Hugo) comment to explain why that replace is happening.

**Comment by sftim** [comment]
Rendered page LGTM

File: layouts/_default/cve-feed.json
Comment (PushkarJ) [reply layouts/_default/cve-feed.json:20]: Done. Hugo did not allow me to add comment in JSON file so added it in the corresponding markdown file

**Comment by sftim** [comment]
Should this PR target the _dev-1.25_ branch?

(I'm not sure - do we want to add the feed with the v1.25 release, or perhaps actually combine the blog announcement for the new feed with the switch-on)

We have options, and several of them are good. I want to be clear on which option we're picking so I can based my approval decision on that context.

**Comment by PushkarJ** [comment]
@sftim I think the #35608 blog being published along with the CVE feed being available by v1.25 sounds like a good idea to me. I have switched target branch to `dev-1.25` now.

**Comment by sftim** [comment]
> @sftim I think the https://github.com/kubernetes/website/pull/35608 blog being published along with the CVE feed being available by v1.25 sounds like a good idea to me

I'm not sure I understand. Which of these is right:
- the feed goes live on v1.25 release day
- the feed goes live after v1.25 release day, simultaneous with the blog announcement about the new feed going live
- something else!

?

**Comment by PushkarJ** [comment]
What I meant was, the feed goes live on v1.25 release day, the feature blog is followed soon after (in next few days). This will allow the feed to have some soak time to iron out any issues we find after going live :)

**Comment by sftim** [comment]
> What I meant was, the feed goes live on v1.25 release day, the feature blog is followed soon after (in next few days). This will allow the feed to have some soak time to iron out any issues we find after going live :)

Sounds good to me.

/approve
/remove-label tide/merge-method-squash

**Comment by sftim** [comment]
(I'd like somebody else to add the LGTM here)

**Comment by PushkarJ** [comment]
/cc @nehaLohia27 @savitharaghunathan 

(Tagging you both for `/lgtm` and/or review comments)

**Comment by katcosgrove** [comment]
Hi from the Comms team! We have you tracked for a feature blog, but the link provided to us was a KEP, not a placeholder blog PR. Are you still intending to have a feature blog for this release?

**Comment by sftim** [comment]
Announcement blog article will be added via https://github.com/kubernetes/website/pull/35608

**Pull request merged**
